### PR TITLE
HDDS-1590 Aggregate bytes write metrics from container to datanode and to SCM

### PR DESCRIPTION
**Pull request by timmylicheng**
## What changes were proposed in this pull request?
#HDDS-1590 Aggregate bytesWrite metrics from container to datanode and all the way to SCM.
(Please fill in changes proposed in this fix)

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-1590

(Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HDDS-XXXX. Fix a typo in YYY.)

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?
UT.
Also tested in env:
TYPE scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_2f0b4887_a6a4_4ab5_ba41_36aa5cfc9056 counter
scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_2f0b4887_a6a4_4ab5_ba41_36aa5cfc9056{hostname="VM_50_210_centos"} 102400

TYPE scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_9b3b4890_b347_4e14_9b84_53d499474473 counter
scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_9b3b4890_b347_4e14_9b84_53d499474473{hostname="VM_50_210_centos"} 307200

TYPE scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_dee39c86_ec69_4edf_bd05_3c5eb16a6ca9 counter
scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_dee39c86_ec69_4edf_bd05_3c5eb16a6ca9{hostname="VM_50_210_centos"} 204800

TYPE scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_398226ba_298d_4b72_9f66_cda4046b037b counter
scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_398226ba_298d_4b72_9f66_cda4046b037b{hostname="VM_50_210_centos"} 11468800

TYPE scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_09c23825_6e8b_4a9d_9d61_603f360c3300 counter
scm_pipeline_metrics_num_pipeline_bytes_write_ratis_three_09c23825_6e8b_4a9d_9d61_603f360c3300{hostname="VM_50_210_centos"} 5

(Please explain how this patch was tested. Ex: unit tests, manual tests)
(If this patch involves UI changes, please attach a screen-shot; otherwise, remove this)

### CODE REVIEW

**Comment by timmylicheng** [comment]
The acceptance details link doesn't work. Wasn't able to see what errors happen. Don't expect this change to affect acceptance test tho.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java [hunk 1]
```diff
@@ -56,11 +56,13 @@
   private @Metric MutableCounterLong numPipelineReportProcessingFailed;
   private @Metric MutableCounterLong numPipelineContainSameDatanodes;
   private Map<PipelineID, MutableCounterLong> numBlocksAllocated;
+  private Map<PipelineID, MutableCounterLong> numPipelineBytesWrite;
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: Change name to numBytesWritten.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java [hunk 2]
```diff
@@ -137,11 +150,16 @@ void incNumPipelineCreated() {
   }
 
   /**
-   * Get the number of pipeline created.
-   * @return number of pipeline
+   * Increments the number of bytes by delta that write into the pipeline.
    */
-  long getNumPipelineCreated() {
-    return numPipelineCreated.value();
+  void incNumPipelineBytesWrite(Pipeline pipeline, long delta) {
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: The reported BytesWritten is delta or sum up?

Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: I have updated the logic here.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java (hunk 1 above)
Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: Updated.

**Comment by timmylicheng** [comment]
> There is no unit test found. Please add corresponding unit test.

In terms of unit test, this change is aggregating bytesWritten stats from containerData to Ratis server and to SCM thru pipeline report. AFAIK, there is no available test framework who goes all the way. I tried MiniOzoneCluster and it didn't contain containerData in its Ratis server. Need to spend more time investigating it.

**Comment by elek** [comment]
Thanks to post this patch @timmylicheng. Overall it looks good, but I am not sure about the motivation.
Why is it required to send metrics from one service to an other one and report from there? The metrics can be published by the datanode itself and can be summarized / organized by the metrics collector.

Is this information required for any of the decision made by the SCM?

I am pretty sure that I missed the context of this patch. Can you please help me to understand the goals?

**Comment by timmylicheng** [comment]
> Thanks to post this patch @timmylicheng. Overall it looks good, but I am not sure about the motivation.
> Why is it required to send metrics from one service to an other one and report from there? The metrics can be published by the datanode itself and can be summarized / organized by the metrics collector.
> 
> Is this information required for any of the decision made by the SCM?
> 
> I am pretty sure that I missed the context of this patch. Can you please help me to understand the goals?

Sure. This patch is to summarize the bytes in for every pipeline, which requires the sum of each container that is allocated to the pipeline. The bytes in stats for each container shall be collected on datanode to be near real-time. So that I collect them up by datanode reports and aggregate by pipeline on SCM.

**Comment by timmylicheng** [comment]
> > Thanks to post this patch @timmylicheng. Overall it looks good, but I am not sure about the motivation.
> > Why is it required to send metrics from one service to an other one and report from there? The metrics can be published by the datanode itself and can be summarized / organized by the metrics collector.
> > Is this information required for any of the decision made by the SCM?
> > I am pretty sure that I missed the context of this patch. Can you please help me to understand the goals?
> 
> Sure. This patch is to summarize the bytes in for every pipeline, which requires the sum of each container that is allocated to the pipeline. The bytes in stats for each container shall be collected on datanode to be near real-time. So that I collect them up by datanode reports and aggregate by pipeline on SCM.

**Comment by timmylicheng** [comment]
> > > Thanks to post this patch @timmylicheng. Overall it looks good, but I am not sure about the motivation.
> > > Why is it required to send metrics from one service to an other one and report from there? The metrics can be published by the datanode itself and can be summarized / organized by the metrics collector.
> > > Is this information required for any of the decision made by the SCM?
> > > I am pretty sure that I missed the context of this patch. Can you please help me to understand the goals?
> > 
> > 
> > Sure. This patch is to summarize the bytes in for every pipeline, which requires the sum of each container that is allocated to the pipeline. The bytes in stats for each container shall be collected on datanode to be near real-time. So that I collect them up by datanode reports and aggregate by pipeline on SCM.

@elek Wonder if you have further questions. :)

**Comment by elek** [comment]
> Sure. This patch is to summarize the bytes in for every pipeline, which requires the sum of each container that is allocated to the pipeline. The bytes in stats for each container shall be collected on datanode to be near real-time. So that I collect them up by datanode reports and aggregate by pipeline on SCM.

Thanks to explain it.

I understand that it's good to calculate the number. I don't understand why is it added to the `StorageContainerDatanodeProtocol.proto` and sent back to the SCM. It seems to be a metric which can be fetched directly from the datanode. We don't need to send it to the SCM.

Unless the SCM needs this data for any **decision** (to create new pipelines, replica containers, etc..). I might missed it but as I see this is not the case (but please let me now if you have such a plan...).

**Comment by timmylicheng** [comment]
> > Sure. This patch is to summarize the bytes in for every pipeline, which requires the sum of each container that is allocated to the pipeline. The bytes in stats for each container shall be collected on datanode to be near real-time. So that I collect them up by datanode reports and aggregate by pipeline on SCM.
> 
> Thanks to explain it.
> 
> I understand that it's good to calculate the number. I don't understand why is it added to the `StorageContainerDatanodeProtocol.proto` and sent back to the SCM. It seems to be a metric which can be fetched directly from the datanode. We don't need to send it to the SCM.
> 
> Unless the SCM needs this data for any **decision** (to create new pipelines, replica containers, etc..). I might missed it but as I see this is not the case (but please let me now if you have such a plan...).

Yea I think you already have the picture. We plan to expose pipeline heavy load in terms of bytes written to make pipeline lifecycle more balanced.

**Comment by timmylicheng** [comment]
> There is no unit test found. Please add corresponding unit test.

@ChenSammi Unit test is added. Please take a look.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java [hunk 3]
```diff
@@ -102,15 +106,24 @@ void createPerPipelineMetrics(Pipeline pipeline) {
...
+        .info(getBytesWriteMetricName(pipeline),
+            "Number of bytes written into pipeline " + pipeline.getId()), 0L));
   }
 
   public static String getBlockAllocationMetricName(Pipeline pipeline) {
     return "NumBlocksAllocated-" + pipeline.getType() + "-" + pipeline
         .getFactor() + "-" + pipeline.getId().getId();
   }
 
+  public static String getBytesWriteMetricName(Pipeline pipeline) {
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: 2

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java [hunk 4]
```diff
@@ -137,11 +150,12 @@ void incNumPipelineCreated() {
   }
 
   /**
-   * Get the number of pipeline created.
-   * @return number of pipeline
+   * Increments the number of bytes by delta that write into the pipeline.
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: delta data or total data?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java [hunk 5]
```diff
@@ -159,21 +159,14 @@ protected void initializePipelineState() throws IOException {
...
-      Pipeline pipeline = pipelineFactory.create(type, factor);
-      pipelineStore.put(pipeline.getId().getProtobuf().toByteArray(),
-          pipeline.getProtobufMessage().toByteArray());
-      stateManager.addPipeline(pipeline);
-      nodeManager.addPipeline(pipeline);
-      metrics.incNumPipelineAllocated();
+  private void recordMetricsForPipeline(Pipeline pipeline) {
+    switch (pipeline.getType()) {
+    case STAND_ALONE:
       if (pipeline.isOpen()) {
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: is this isOpen check duplicated?

File: hadoop-hdds/container-service/src/main/proto/StorageContainerDatanodeProtocol.proto [hunk 6]
```diff
@@ -240,6 +240,7 @@ message ContainerAction {
 message PipelineReport {
   required PipelineID pipelineID = 1;
   required bool isLeader = 2;
+  optional uint64 bytesWrite = 3;
```
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-hdds/container-service/src/main/proto/StorageContainerDatanodeProtocol.proto]: Updated.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/XceiverServerRatis.java [hunk 7]
```diff
@@ -579,15 +581,33 @@ public boolean isExist(HddsProtos.PipelineID pipelineId) {
         RaftGroupId.valueOf(PipelineID.getFromProtobuf(pipelineId).getId()));
   }
 
+  private long calculateBytesWriteForRelatedPipeline(
```
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/XceiverServerRatis.java]: Updated.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java [hunk 8]
```diff
@@ -104,6 +106,12 @@ protected void processPipelineReport(PipelineReport report,
 
     setReportedDatanode(pipeline, dn);
     setPipelineLeaderId(report, pipeline, dn);
+    // ONE replica pipeline doesn't have leader flag
+    if (report.getIsLeader() ||
+        pipeline.getFactor() == HddsProtos.ReplicationFactor.ONE) {
+      pipeline.setLeaderId(dn.getUuid());
+      metrics.incNumPipelineBytesWrite(pipeline, report.getBytesWrite());
```
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java]: Updated.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java (hunk 3 above)
Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: Updated.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java (hunk 5 above)
Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: Works as a kind of protection. Current way to create STAND_ALONE pipeline is to put the state as OPEN indeed. But it's not guaranteed.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java (hunk 4 above)
Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineMetrics.java]: Update to total

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java (hunk 5 above)
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: In this case， metrics.createPerPipelineMetrics(pipeline)  is executed twice for standalone pipeline.

Comment (timmylicheng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: Updated

**Comment by ChenSammi** [comment]
+1.  Will commit it shortly.

**Pull request merged**